File: 2025/4.js
const input = [
  [".", ".", "@", "@", ".", "@", "@", "@", "@", "."],
  ["@", "@", "@", ".", "@", ".", "@", ".", "@", "@"],
  ["@", "@", "@", "@", "@", ".", "@", ".", "@", "@"],
  ["@", ".", "@", "@", "@", "@", ".", ".", "@", "."],
  ["@", "@", ".", "@", "@", "@", "@", ".", "@", "@"],
  [".", "@", "@", "@", "@", "@", "@", "@", ".", "@"],
  [".", "@", ".", "@", ".", "@", ".", "@", "@", "@"],
  ["@", ".", "@", "@", "@", ".", "@", "@", "@", "@"],
  [".", "@", "@", "@", "@", "@", "@", "@", "@", "."],
  ["@", ".", "@", ".", "@", "@", "@", ".", "@", "."],
];

function countAdjacentRolls(input, i, j) {
  let adjacentRolls = 0;

  const adjacentCoordinates = [
    { i: 0, j: 1 },
    { i: 1, j: 1 },
    { i: 1, j: 0 },
    { i: 1, j: -1 },
    { i: 0, j: -1 },
    { i: -1, j: -1 },
    { i: -1, j: 0 },
    { i: -1, j: 1 },
  ];

  for (const coordinate of adjacentCoordinates) {
    if (input?.[i + coordinate.i]?.[j + coordinate.j] === "@") adjacentRolls++;
  }

  return adjacentRolls;
}

let validRolls = 0;

for (let i = 0; i < input.length; i++) {
  for (let j = 0; j < input[i].length; j++) {
    if (input[i][j] === ".") continue;

    const adjacentRolls = countAdjacentRolls(input, i, j);

    if (adjacentRolls < 4) validRolls++;
  }
}

console.log(validRolls);

File: 2025/4.go
package main

import (
	"fmt"
)

type coordinate struct {
	i, j int
}

func countAdjacentRolls(input [][]string, i, j int) int {
	adjacentRolls := 0

	adjacentCoordinates := [...]coordinate{
		{i: 0, j: 1},
		{i: 1, j: 1},
		{i: 1, j: 0},
		{i: 1, j: -1},
		{i: 0, j: -1},
		{i: -1, j: -1},
		{i: -1, j: 0},
		{i: -1, j: 1},
	}

	for _, coordinate := range adjacentCoordinates {
		if i+coordinate.i < 0 {
			continue
		}

		if j+coordinate.j < 0 {
			continue
		}

		if i+coordinate.i >= len(input) {
			continue
		}

		if j+coordinate.j >= len(input[0]) {
			continue
		}

		if input[i+coordinate.i][j+coordinate.j] == "@" {
			adjacentRolls++
		}
	}

	return adjacentRolls
}

func main() {
	input := [...][]string{
		{".", ".", "@", "@", ".", "@", "@", "@", "@", "."},
		{"@", "@", "@", ".", "@", ".", "@", ".", "@", "@"},
		{"@", "@", "@", "@", "@", ".", "@", ".", "@", "@"},
		{"@", ".", "@", "@", "@", "@", ".", ".", "@", "."},
		{"@", "@", ".", "@", "@", "@", "@", ".", "@", "@"},
		{".", "@", "@", "@", "@", "@", "@", "@", ".", "@"},
		{".", "@", ".", "@", ".", "@", ".", "@", "@", "@"},
		{"@", ".", "@", "@", "@", ".", "@", "@", "@", "@"},
		{".", "@", "@", "@", "@", "@", "@", "@", "@", "."},
		{"@", ".", "@", ".", "@", "@", "@", ".", "@", "."},
	}

	validRolls := 0

	for i := 0; i < len(input); i++ {
		for j := 0; j < len(input[0]); j++ {
			if input[i][j] == "." {
				continue
			}

			adjacentRolls := countAdjacentRolls(input[0:len(input)], i, j)

			if adjacentRolls < 4 {
				validRolls++
			}
		}

	}

	fmt.Println(validRolls)

}
